### PR TITLE
Add Network ACLs creation feature

### PR DESCRIPTION
**Pull request by s2504s**
## What 
* Added the ability to create `Network ACLs`
* Removed `null_resource` for cidrsubnet
* Removed `aws_internet_gateway` datasource

## Why
* We need this in order to consolidate network security rules to security groups only
* `null_resource.triggers`  is not iterable
```
data.aws_availability_zones.available: Refreshing state...
Error running plan: 2 error(s) occurred:
* module.vpc.module.subnets.aws_subnet.private: 2 error(s) occurred:
* module.vpc.module.subnets.aws_subnet.private[0]: Resource 'null_resource.private_cidr_block' does not have attribute 'triggers.subnet' for variable 'null_resource.private_cidr_block.*.triggers.subnet'
* module.vpc.module.subnets.aws_subnet.private[1]: Resource 'null_resource.private_cidr_block' does not have attribute 'triggers.subnet' for variable 'null_resource.private_cidr_block.*.triggers.subnet'
* module.vpc.module.subnets.aws_subnet.public: 2 error(s) occurred:
* module.vpc.module.subnets.aws_subnet.public[0]: Resource 'null_resource.public_cidr_block' does not have attribute 'triggers.subnet' for variable 'null_resource.public_cidr_block.*.triggers.subnet'
* module.vpc.module.subnets.aws_subnet.public[1]: Resource 'null_resource.public_cidr_block' does not have attribute 'triggers.subnet' for variable 'null_resource.public_cidr_block.*.triggers.subnet'
```
* TF modules don't have `depends_on` capability. But `tf_subnets` should depend on `aws_internet_gateway` resource of [tf_vpc](https://github.com/cloudposse/tf_vpc)

### CODE REVIEW

File: acls.tf [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  tags       = "${var.tags}"
+}
+
+resource "aws_network_acl" "default" {
+  count      = "${var.create_network_acl ? 1 : 0}"
+  vpc_id     = "${var.vpc_id}"
+  subnet_ids = ["${compact(concat(aws_subnet.public.*.id, aws_subnet.private.*.id))}"]
+
+  egress {
+    rule_no    = 100
```
Comment (osterman) [inline acls.tf]: Have you confirmed this is before the default rule which is to deny everything?

File: main.tf [hunk 2]
```diff
@@ -13,10 +13,17 @@ data "aws_vpc" "default" {
 
 data "aws_availability_zones" "available" {}
 
+resource "null_resource" "default" {
+  triggers = {
+    default = "${join(" ", var.depends_on)}"
```
Comment (osterman) [inline main.tf]: I would call `default` as `depends_on`.

File: main.tf [hunk 3]
```diff
@@ -13,10 +13,17 @@ data "aws_vpc" "default" {
 
 data "aws_availability_zones" "available" {}
 
+resource "null_resource" "default" {
```
Comment (osterman) [inline main.tf]: I don't understand why this `null_resource` is necessary.

File: acls.tf [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  source     = "git::https://github.com/cloudposse/tf_label.git?ref=tags/0.2.0"
+  namespace  = "${var.namespace}"
+  stage      = "${var.stage}"
+  name       = "${var.name}"
+  delimiter  = "${var.delimiter}"
+  attributes = "${var.attributes}"
+  tags       = "${var.tags}"
+}
+
+resource "aws_network_acl" "default" {
```
Comment (osterman) [inline acls.tf]: Create a public and private network acl.

File: acls.tf [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  ingress {
+    rule_no    = 100
+    action     = "allow"
+    cidr_block = "0.0.0.0/0"
+    from_port  = 0
+    to_port    = 0
+    protocol   = "-1"
+  }
+
+  tags = "${module.acl_label.tags}"
```
Comment (osterman) [inline acls.tf]: and public/private label.

File: README.md [hunk 6]
```diff
@@ -16,7 +16,9 @@ you plan to use new (separate) VPC.
 * `namespace`: (Required) Namespace associated with these resources
 * `vpc_id`: (Required) AWS Virtual Private Cloud ID.
 * `vpc_default_route_table_id`: A default route table for public subnets. Provides access to Internet. If not set here - will be created.
+* `create_network_acl`: The new set of ACLs that will be associated with these resources
```
Comment (osterman) [inline README.md]: I think we should always create it unless the user specifies a `public_network_acl` or `private_network_acl`

File: README.md [hunk 7]
```diff
@@ -29,6 +31,30 @@ module "subnets" {
...
+  availability_zones         = "${var.availability_zones}"
+  namespace                  = "${var.namespace}"
+  name                       = "${var.name}"
+  stage                      = "${var.stage}"
+  region                     = "${var.region}"
+  vpc_id                     = "${data.aws_vpc.default.id}"
+  cidr_block                 = "${data.aws_vpc.default.cidr_block}"
+  vpc_default_route_table_id = "${var.vpc_default_route_table_id}"
+  create_network_acl         = true
+  depends_on                 = ["${module.vpc.internet_gateway_id}"]
```
Comment (osterman) [inline README.md]: This example is incongruent with the `data.aws_vpc.default`

File: acls.tf (hunk 1 above)
Comment (s2504s) [reply acls.tf]: Yes. The `Deny ALL` rule always is on the last line

File: main.tf (hunk 3 above)
Comment (s2504s) [reply main.tf]: We get the existing `igw` by datasource `data.aws_internet_gateway.default` in the file `main.tf`
If there is the resource `igw` is not available then `terraform apply` is failed with error:
```![image](https://user-images.githubusercontent.com/1134449/30314598-dc0e3b46-97a9-11e7-99d8-e58413e0b952.png)

File: main.tf [hunk 8]
```diff
@@ -13,10 +13,17 @@ data "aws_vpc" "default" {
 
 data "aws_availability_zones" "available" {}
 
+resource "null_resource" "depends_on" {
```
Comment (comeanother) [inline main.tf]: If we return `igw_id` (see #11), this resource may be not required.

File: variables.tf [hunk 9]
```diff
@@ -1,31 +1,50 @@
...
 }
 
-variable "namespace" {
+variable "vpc_default_route_table_id" {
   default = ""
 }
 
-variable "stage" {
-  default = ""
+variable "create_network_acl" {
```
Comment (comeanother) [inline variables.tf]: This variable seems to be unused. Drop it.

File: main.tf (hunk 8 above)
Comment (s2504s) [reply main.tf]: Returned

File: acls.tf [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+  source     = "git::https://github.com/cloudposse/tf_label.git?ref=tags/0.2.0"
+  namespace  = "${var.namespace}"
+  stage      = "${var.stage}"
+  name       = "${var.name}"
+  delimiter  = "${var.delimiter}"
+  attributes = ["private"]
+  tags       = "${var.tags}"
+}
+
+resource "aws_network_acl" "public" {
```
Comment (osterman) [inline acls.tf]: Move public to `public.tf`

File: acls.tf [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+    cidr_block = "0.0.0.0/0"
+    from_port  = 0
+    to_port    = 0
+    protocol   = "-1"
+  }
+
+  tags = "${module.acl_public_label.tags}"
+}
+
+resource "aws_network_acl" "private" {
```
Comment (osterman) [inline acls.tf]: Move to `private.tf`

Comment (osterman) [reply acls.tf]: ...and corresponding dependencies.

File: acls.tf (hunk 10 above)
Comment (osterman) [reply acls.tf]: ...and corresponding dependencies.

File: private.tf [hunk 12]
```diff
@@ -5,25 +5,12 @@ module "private_label" {
   name      = "${var.name}-private"
 }
 
-resource "null_resource" "private_cidr_block" {
```
Comment (osterman) [inline private.tf]: Why couldn't we just have a `depends_on` here for the vpc?

File: public.tf [hunk 13]
```diff
@@ -5,25 +5,12 @@ module "public_label" {
   name      = "${var.name}-public"
 }
 
-resource "null_resource" "public_cidr_block" {
```
Comment (osterman) [inline public.tf]: Why do we need to get rid of the `null_resource`? It seems cleaner to me.

**Comment by comeanother** [comment]
Tested.
screenshot from 2017-09-13 21-25-56

**Pull request merged**